contracts/cw721-base/src/contract.rs

```diff
@@ -14,442 +12,609 @@ use cw721::{
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    fn default() -> Self {
+        Self::new(
+            "nft_info",
+            "minter",
+            "num_tokens",
+            "operators",
+            "tokens",
+            "tokens__owner",
+        )
     }
 }
 
-pub fn execute_mint(
-    deps: DepsMut,
-    _env: Env,
-    info: MessageInfo,
-    msg: MintMsg,
-) -> Result<Response, ContractError> {
-    let minter = MINTER.load(deps.storage)?;
+impl<'a, T, C> Cw721Contract<'a, T, C>
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    fn new(
+        contract_key: &'a str,
+        minter_key: &'a str,
+        token_count_key: &'a str,
+        operator_key: &'a str,
+        tokens_key: &'a str,
+        tokens_owner_key: &'a str,
+    ) -> Self {
+        let indexes = TokenIndexes {
+            owner: MultiIndex::new(token_owner_idx, tokens_key, tokens_owner_key),
+        };
+        Self {
+            contract_info: Item::new(contract_key),
+            minter: Item::new(minter_key),
+            token_count: Item::new(token_count_key),
+            operators: Map::new(operator_key),
+            tokens: IndexedMap::new(tokens_key, indexes),
+            _custom_response: PhantomData,
+        }
+    }
 
-    if info.sender != minter {
-        return Err(ContractError::Unauthorized {});
+    pub fn num_tokens(&self, storage: &dyn Storage) -> StdResult<u64> {
+        Ok(self.token_count.may_load(storage)?.unwrap_or_default())
     }
 
-    // create the token
-    let token = TokenInfo {
-        owner: deps.api.addr_validate(&msg.owner)?,
-        approvals: vec![],
-        name: msg.name,
-        description: msg.description.unwrap_or_default(),
-        image: msg.image,
-    };
-    tokens().update(deps.storage, &msg.token_id, |old| match old {
-        Some(_) => Err(ContractError::Claimed {}),
-        None => Ok(token),
-    })?;
-
-    increment_tokens(deps.storage)?;
-
-    Ok(Response::new()
-        .add_attribute("action", "mint")
-        .add_attribute("minter", info.sender)
-        .add_attribute("token_id", msg.token_id))
+    pub fn increment_tokens(&self, storage: &mut dyn Storage) -> StdResult<u64> {
+        let val = self.num_tokens(storage)? + 1;
+        self.token_count.save(storage, &val)?;
+        Ok(val)
+    }
 }
 
-pub fn execute_transfer_nft(
-    deps: DepsMut,
-    env: Env,
-    info: MessageInfo,
-    recipient: String,
-    token_id: String,
-) -> Result<Response, ContractError> {
-    _transfer_nft(deps, &env, &info, &recipient, &token_id)?;
-
-    Ok(Response::new()
-        .add_attribute("action", "transfer_nft")
-        .add_attribute("sender", info.sender)
-        .add_attribute("recipient", recipient)
-        .add_attribute("token_id", token_id))
-}
+impl<'a, T, C> Cw721Contract<'a, T, C>
+where
+    T: Serialize + DeserializeOwned + Clone,
+    C: Clone + std::fmt::Debug + PartialEq + JsonSchema,
+{
+    pub fn instantiate(
```

Comment:
In contradiction to `execute`/`query` this should stay in contract itself (as it is very much per contract stuff).